cmd/vault-login-oauth/vault-login-oauth.go

```diff
@@ -0,0 +1,227 @@
+	"encoding/base64"
+	"errors"
+	"flag"
+	"fmt"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"runtime"
+	"time"
+
+	"github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/command/token"
+)
+
+type oidcLogin struct {
+	server       *http.Server
+	listener     net.Listener
+	vault        *api.Logical
+	authReqURL   string
+	redirectURL  string
+	stateNonce   string
+	authDone     context.CancelFunc
+	authErr      error
+	authResponse *api.Secret
+	role         string
+}
+
+func newLogin(pluginPath string, role string) *oidcLogin {
+	login := &oidcLogin{
+		server: &http.Server{},
+		role:   role,
+	}
+
+	// The callback URI must be unique for each authorization server. Since
+	// each mount of the plugin corresponds to an authorization server,
+	// incorporating the pluginPath accomplishes that.
+	//
+	// https://tools.ietf.org/html/rfc8252#section-8.10
+	callbackPath := pluginPath + "/cb"
+
+	// Create our vault client.
+	{
+		vaultClient, err := api.NewClient(nil)
+		if err != nil {
+			fatal("Couldn't create Vault client:", err)
+		}
+		login.vault = vaultClient.Logical()
+	}
+
+	// Start listening on some random localhost port, at which point we'll know
+	// what redirect_uri to give the authorization server.
+	{
+		listener, err := net.Listen("tcp", "127.0.0.1:0")
+		if err != nil {
+			fatal(err)
+		}
+		login.listener = listener
+		login.redirectURL = fmt.Sprintf("http://%s%s", login.listener.Addr(), callbackPath)
+	}
+
+	// Determine the authorization request URL.
+	{
+		secret, err := login.vault.Read(pluginPath + "/auth-request")
+		if err != nil || secret == nil || secret.Data == nil {
+			fatal("Failed to get authorization request URL:", err)
+		}
+		rawURL := secret.Data["url"].(string)
+		parsedURL, err := url.Parse(rawURL)
+		if err != nil {
+			fatal("Failed parsing authorization request URL from Vault:", err)
+		}
+
+		login.stateNonce = generateNonce()
+
+		query := parsedURL.Query()
+		query.Set("redirect_uri", login.redirectURL)
+		query.Set("state", login.stateNonce)
+		parsedURL.RawQuery = query.Encode()
+		login.authReqURL = parsedURL.String()
+	}
+
+	http.HandleFunc(callbackPath, login.handleOAuthCallback)
+
+	return login
+}
+
+// HTTP handler for the redirect_uri. Once the user has gone through the
+// authentication flow at the identity provider they will be redirected here,
+// at which point we obtain the authorization code if authentication was
+// successful, or an error otherwise.
+func (login *oidcLogin) handleOAuthCallback(writer http.ResponseWriter, req *http.Request) {
+	writer.Header().Set("Content-Type", "text/plain; charset=utf-8")
+
+	query := req.URL.Query()
+	code := query.Get("code")
+	state := query.Get("state")
+
+	if state != login.stateNonce {
+		// This indicates either a CSRF attack or a buggy authorization server.
+		// We won't even dignify it with a response.
+		fatal("DANGER: POSSIBLE CSRF ATTACK DETECTED.\nReceived state", state, "expected", login.stateNonce)
```

Comment:
State is recommended, but not required. Might want to accept or allow for configuration to specify nil is ok.

Reply:
In every case, error recovery in this program involves just exiting. I thought rather than write two dozen times `if err != nil { return nil, err }` I might as well just cut out the boilerplate and get straight to business. Is that weird?